Added SIP-3 (#29)
MetaMask/SIPs: SIPS/sip-3.md

---
sip: 3
title: Transaction Insights
status: Review
discussions-to: https://github.com/MetaMask/SIPs/discussions/31

> author: Hassan Malik (@hmalik88)

Author: Naturally, sir 😄

created: 2022-08-23
---

## Table of Contents

- [Table of Contents](#table-of-contents)
- [Abstract](#abstract)
- [Motivation](#motivation)
- [Specification](#specification)
  - [Language](#language)
  - [Definitions](#definitions)
  - [Common types](#common-types)
  - [Snap Developer](#snap-developer)
    - [Manifest](#manifest)
    - [Snap](#snap)
- [History](#history)
- [Copyright](#copyright)

## Abstract

This SIP proposes a way for snaps to provide transaction insights that can then be displayed in the MetaMask confirmation UI.

> Example use-cases for transaction insights can be phishing detection, malicious address detection, transaction simulation, arbitrage trading.

Reviewer: ```suggestion
Example use-cases for transaction insights are phishing detection, malicious contract detection, transaction simulation and arbitrage trading.
```

> Example use-cases for transaction insights can be phishing detection, malicious address detection, transaction simulation, arbitrage trading.

Author: Ok we can take that out, I was thinking a long the lines of a trading bot snap using `onTransaction` to get insights from which they can decide to make a trade or not. BUT that doesn't really make sense considering that is only ever triggered from the confirmation window 😅

## Motivation

> One of the main use cases for transaction insights is transaction simulation. In the current state of the MetaMask wallet, we do not provide any sort of insight as to whether or not a transaction will succeed based on the web of contract/address interaction that can happen. Transaction insight snaps would provide the ability to calculate some sort of insights and return them to the wallet..

Reviewer: ```suggestion
Deciding what information to show before a user is prompted to sign a transaction and furthermore deciding which information is critical is a difficult problem. It may even be user-dependant. Similarly to how SIP-2 allows snaps to expand the keyrings that MetaMask support, this SIP aims to expand the options the user is given in which information they see before signing a transaction.

The current MetaMask extension already has a "transaction insights" feature that does transaction simulation and shows the result to the user. This SIP aims to expand on this feature, allowing the community to build snaps that provide **any form of "insight"** into a transaction. This insight will be shown in the MetaMask UI alongside any official insight provided by MetaMask itself.
```

Rewrote this a bit. Let me know what you think!

Author: Looks good!

## Specification

> > Formal specifications are written in Typescript and [JSON schema version 2020-12](https://json-schema.org/draft/2020-12/json-schema-core.html). Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs)

Reviewer 2: I don't see any JSON schema specifications in this SIP

> Formal specifications are written in Typescript and [JSON schema version 2020-12](https://json-schema.org/draft/2020-12/json-schema-core.html). Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs)

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### Definitions

> This section is non-normative

- `ChainId` - a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) string. It identifies a specific chain in all of possible blockchains.
  - `ChainId` consists of a `Namespace` and a `Reference`
    - `Namespace` - A class of similar blockchains. For example EVM-based blockchains.
    - `Reference` - A way to identify a concrete chain inside a `Namespace`. For example Ethereum Mainnet or Polygon.

> - `AccountId` - a [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md) string. It identifies a specific account on a specific chain.

Reviewer: Not referenced anywhere

> - `AccountId` - a [CAIP-10](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-10.md) string. It identifies a specific account on a specific chain.

Author: Left in by accident

### Common types

The below common types are used throughout the specification

````typescript
type ChainId = string;

- `ChainId` strings MUST be [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) Chain Id.

  The Regular Expression used to validate Chain IDs by the snap SHOULD be:

>   ```typescript

Reviewer 2: The regex was added to SIP-2 because the wallet has to validate the data from the snap. But in case of this SIP, there shouldn't be any need for validation in usage of this API

Author: The `chainId` in the `onTransaction` handler function's params is CAIP-2, can't the snap developer use this regex to extract the desired chainId and use that to get whatever insight?

Author: Ah I see, okay that makes sense 😄 -- I'll remove it.

An example usage of the permission inside `snap.manifest.json` is as follows:

```json
{
  "initialPermissions": {
    "endowment:transaction-insight": {}
  }
}
```

#### Snap
  
The transaction insight snap implementation must implement the following API:

```typescript
import { OnTransactionHandler } from '@metamask/snap-types';

export const onTransaction: OnTransactionHandler = async ({ transaction, chainId }) => {
    // do something
    return { insights }
}
```

The interface for an `onTransaction` handler function’s arguments is:

```typescript
interface OnTransactionArgs  {
    transaction: Record<string, unknown>;
    chainId: string;
}
```

**Transaction** - The transaction object is specifically not defined in this SIP because the transaction object can look different across various chains and it is not our intention to define an interface for every chain. Instead, the onus is on the Snap developer to be cognizant of the shape of the transaction object. However, that being said, the transaction object out of the MetaMask extension is of the following interface:

*[NON EIP-1559]*

```typescript

> interface TransactionObject {

Reviewer: This was not applied, but solved. Can we discuss if you disagree?

> interface TransactionObject {

Author: It got reversed when I pulled in Olaf's commits, fixing it now.

    from: string;
    to: string;
    nonce: string;
    value: string;
    data: string;
    gas: string;
    gasPrice: string;

>     Type: string;

Author: `Type`-o lol

    estimateSuggested: string;
    estimateUsed: string;
}
```

*[EIP-1559]*

```typescript
interface TransactionObject {
	from: string;
	to: string;
	nonce: string;
	value: string;
	data: string;
	gas: string;

> 	maxFeePerrGas: string;

Reviewer: This has not been solved yet

> 	maxFeePerrGas: string;

Author: Same as above^

  gas: string;
  maxFeePerrGas: string;
  maxPriorityFeePerGas: string;
  Type: string;
  estimateSuggested: string;
  estimateUsed: string;
}
```

**ChainId** - This is a CAIP-2 `ChainId` string, the snap is expected to parse and utilize this string as needed.

The return object for an `onTransaction` import should be as follows:

```typescript
interface OnTransactionReturn {

>   insights: Record<string, unknown>;

Author: Ok will add some explanation for this.

}
```

## History

> The transaction insight feature has been inspired by the need to improve user confidence in everyday web3 interactions. Allowing for various transaction insights inside of MetaMask will increase confidence, improve UX and ultimately drive the volume in transactions across our user base.

Reviewer: ```suggestion
The transaction insight feature has been inspired by the need to improve user confidence in everyday web3 interactions. Allowing for various transaction insights inside of MetaMask will increase confidence, improve UX and ultimately provide a safer and more informed experience for our users.
```

Reviewer: Or maybe just kill this part entirely?

## Specification

> Formal specifications are written in Typescript and [JSON schema version 2020-12](https://json-schema.org/draft/2020-12/json-schema-core.html). Usage of `CAIP-N` specifications, where `N` is a number, are references to [Chain Agnostic Improvement Proposals](https://github.com/ChainAgnostic/CAIPs)

### Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)

### Definitions

> This section is non-normative

- `ChainId` - a [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) string. It identifies a specific chain in all of possible blockchains.
  - `ChainId` consists of a `Namespace` and a `Reference`
    - `Namespace` - A class of similar blockchains. For example EVM-based blockchains.
    - `Reference` - A way to identify a concrete chain inside a `Namespace`. For example Ethereum Mainnet or Polygon.

### Common types

The below common types are used throughout the specification

````typescript

> type ChainId = string;

Reviewer 2: Unclosed code tag

- `ChainId` strings MUST be [CAIP-2](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-2.md) Chain Id.

  The Regular Expression used to validate Chain IDs by the snap SHOULD be:

>   ```typescript

Reviewer 2: Tabs before code tag

  const chainIdValidation =
    /^(?<namespace>[-a-z0-9]{3,8}):(?<reference>[-a-zA-Z0-9]{1,32})$/;

> ````

Reviewer 2: Too many apostrophes for the end code tag
